packages/mana/manabasevector.go

```diff
@@ -175,6 +164,75 @@ func (m *ManaBaseVector) Book(txInfo *TxInfo) {
 	}
 }
 
+func (m *ManaBaseVector) BookEpoch(created []*ledger.OutputWithMetadata, spent []*ledger.OutputWithMetadata) {
+	var revokeEvents []*RevokedEvent
+	var pledgeEvents []*PledgedEvent
+	var updateEvents []*UpdatedEvent
+	// only lock mana vector while we are working with it
+	func() {
+		m.Lock()
+		defer m.Unlock()
+		// first, revoke mana from previous owners
+		for _, output := range spent {
+			idToRevoke := m.getIDBasedOnManaType(output)
+			outputIOTAs, _ := output.Output().(devnetvm.Output).Balances().Get(devnetvm.ColorIOTA)
+			oldMana := m.getOldManaAndRevoke(idToRevoke, float64(outputIOTAs))
+			// save events for later triggering
+			revokeEvents = append(revokeEvents, &RevokedEvent{idToRevoke, float64(outputIOTAs), output.OutputMetadata().CreationTime(), m.Type(), output.ID().TransactionID, output.ID()})
+			updateEvents = append(updateEvents, &UpdatedEvent{idToRevoke, &oldMana, m.M.Vector[idToRevoke], m.Type()})
+
+		}
+		// second, pledge mana to new nodes
+		for _, output := range created {
+			idToPledge := m.getIDBasedOnManaType(output)
+			outputIOTAs, _ := output.Output().(devnetvm.Output).Balances().Get(devnetvm.ColorIOTA)
+			oldMana := m.getOldManaAndPledge(idToPledge, float64(outputIOTAs))
+			pledgeEvents = append(pledgeEvents, &PledgedEvent{idToPledge, float64(outputIOTAs), output.OutputMetadata().CreationTime(), m.Type(), output.Output().ID().TransactionID})
+			updateEvents = append(updateEvents, &UpdatedEvent{idToPledge, &oldMana, m.M.Vector[idToPledge], m.Type()})
+		}
+	}()
+
+	m.triggerManaEvents(revokeEvents, pledgeEvents, updateEvents)
+}
+
+func (m *ManaBaseVector) getIDBasedOnManaType(output *ledger.OutputWithMetadata) identity.ID {
+	var id identity.ID
+	if m.Type() == ConsensusMana {
+		id = output.OutputMetadata().ConsensusManaPledgeID()
+	} else {
+		id = output.OutputMetadata().AccessManaPledgeID()
+	}
+	return id
+}
+
+func (m *ManaBaseVector) getOldManaAndRevoke(oldPledgeNodeID identity.ID, amount float64) ManaBase {
+	if _, exist := m.M.Vector[oldPledgeNodeID]; !exist {
+		// first time we see this node
+		m.M.Vector[oldPledgeNodeID] = &ManaBase{}
+		m.M.Vector[oldPledgeNodeID].Init()
+	}
+	// save old mana
+	oldMana := *m.M.Vector[oldPledgeNodeID]
+	// revoke BM1
+	err := m.M.Vector[oldPledgeNodeID].revoke(amount)
+	if errors.Is(err, ErrBaseManaNegative) {
```

Comment:
Just checking: if it is the first time we saw this node this error will trigger right? I think it is ok that it produces a panic, it should never happen that you revoke some Mana you don't even know the existence of.

Reply:
Not sure if it will always be the first time we see the nodeID as those are inputs. Previously we had panic but we booked tx by tx, so it should not happen that the balance will be negative. However with BookEpoch it should be checked only after all diffs are processed. I should change it